enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java

```diff
@@ -163,6 +168,51 @@ private KernelTransaction getCurrentTx()
     ==================================================================================
      */
 
+    @Description( "List all procedures in the DBMS." )
+    @Procedure( name = "dbms.procedures", mode = DBMS )
+    public Stream<ProcedureResult> listProcedures()
+    {
+        return graph.getDependencyResolver().resolveDependency( Procedures.class ).getAllProcedures().stream()
+                .sorted( ( a, b ) -> a.name().toString().compareTo( b.name().toString() ) )
+                .map( sig -> new ProcedureResult( sig ) );
+    }
+
+    @SuppressWarnings( "WeakerAccess" )
+    public static class ProcedureResult
+    {
+        public final String name;
+        public final String signature;
+        public final String description;
+        public final List<String> roles;
+
+        public ProcedureResult( ProcedureSignature signature )
+        {
+            this.name = signature.name().toString();
+            this.signature = signature.toString();
+            this.description = signature.description().orElse( "" );
+            roles = new ArrayList<>();
+            switch ( signature.mode() )
+            {
+            case DBMS:
+                roles.add( "admin" );
+                break;
+            case READ_ONLY:
+                roles.add( "reader" );
+            case READ_WRITE:
+                roles.add( "publisher" );
+            case SCHEMA_WRITE:
+                roles.add( "architect" );
```

Comment:
So will this switch add the architect role to a `mode=READ` query? Do we want that?